### PR TITLE
Support for account deletion in ledgers

### LINKED ISSUE #15742: Add deletion functionality in ledger implementations
Add a `remove` function in  ledger implementations. 

The actual implementation will need to keep track of removed locations that are available for further account allocations (aka creations) through a stack-based data structure.

The main technical subtlety lies in soundly updating hashes for data structures.

Acceptance criteria (basically testing)
- tests cover nominal behaviors and corner cases
- coverage is XX % (all the lines that use the new data structure are covered?)
- if there are no tests, this shouldn't be acceptable

### PR DESCRIPTION
**Pull request by rbonichon**
This is *draft* PR, intended to show and track implementation progress.

The commits at the moment are mostly informative and **will be rebased** to get a cleaner history, once it is ready. 

This fixes #15742

### CODE REVIEW

**Comment by svv232** [comment]
looks interesting so far! i am curious to see how the ledger will eventually be affected

File: src/lib/merkle_ledger/intf.ml [hunk 1]
```diff
@@ -332,18 +332,16 @@ module Ledger = struct
       t -> init:'accum -> f:(Addr.t -> 'accum -> account -> 'accum) -> 'accum
 
     (** the set of [account_id]s are ledger elements to skip during the fold,
-      because they're in a mask
-  *)
+        because they're in a mask *)
```
Comment (mrmr1993) [inline src/lib/merkle_ledger/intf.ml:335]: Prefer closing doc-comment on new line, per the style guide.

Comment (rbonichon) [reply src/lib/merkle_ledger/intf.ml]: Done

File: src/app/archive/lib/extensional.ml [hunk 2]
```diff
@@ -116,7 +116,7 @@ module Block = struct
   module Stable = struct
     [@@@with_versioned_json]
 
-    module V2 = struct
+    module V3 = struct
```
Comment (mrmr1993) [inline src/app/archive/lib/extensional.ml:119]: This (and similar changes) are not 'compatible', and as such can't be landed in the `compatible` branch.

Comment (rbonichon) [reply src/app/archive/lib/extensional.ml]: This is fixed. This work is in #15832
